### PR TITLE
Public API Strong Typing for kiwi package_manager

### PR DESCRIPTION
**Pull request by KeithMnemonic**
Apply Strong Typing to KIWI Public APIs for package_manager files

 package_manager/base.py
 package_manager/pacman.py
 package_manager/microdnf.py
 package_manager/apt.py
 package_manager/dnf.py
 package_manager/zypper.py

This references Issue #1644

### CODE REVIEW

File: kiwi/package_manager/base.py [hunk 1]
```diff
@@ -28,17 +32,18 @@ class PackageManagerBase:
...
-        self.exclude_requests = []
+        self.package_requests: List[str] = []
+        self.collection_requests: List[str] = []
+        self.product_requests: List[str] = []
+        self.exclude_requests: List[str] = []
 
         self.post_init(custom_args)
 
-    def post_init(self, custom_args=None):
+    def post_init(self, custom_args: list = None) -> None:
```
Comment (schaefi) [inline kiwi/package_manager/base.py:46]: Should this not also use the List type from Typing ?

File: kiwi/package_manager/apt.py [hunk 2]
```diff
@@ -41,7 +43,8 @@ class PackageManagerApt(PackageManagerBase):
     :param dict command_env: apt-get command environment from repository
         runtime configuration
     """
-    def post_init(self, custom_args=None):
+
```
Comment (schaefi) [inline kiwi/package_manager/apt.py:46]: Any reason for the extra empty line ?

File: kiwi/package_manager/base.py [hunk 3]
```diff
@@ -124,73 +129,82 @@ def process_delete_requests(self, force=False):
...
+    def process_plus_recommended(self) -> None:
         """
         Setup package processing to also include recommended dependencies
 
         Implementation in specialized package manager class
         """
         raise NotImplementedError
 
-    def match_package_installed(self, package_list, log_line):
+    def match_package_installed(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/base.py:156]: list vs List ?

File: kiwi/package_manager/dnf.py [hunk 4]
```diff
@@ -36,7 +38,7 @@ class PackageManagerDnf(PackageManagerBase):
     :param dict command_env: dnf command environment from repository runtime
         configuration
     """
-    def post_init(self, custom_args=None):
+    def post_init(self, custom_args: list = None) -> None:
```
Comment (schaefi) [inline kiwi/package_manager/dnf.py:41]: list vs. List

File: kiwi/package_manager/dnf.py [hunk 5]
```diff
@@ -211,21 +213,21 @@ def update(self):
...
-    def process_plus_recommended(self):
+    def process_plus_recommended(self) -> None:
         """
         Setup package processing to also include recommended dependencies.
         """
         if '--setopt=install_weak_deps=False' in self.custom_args:
             self.custom_args.remove('--setopt=install_weak_deps=False')
 
-    def match_package_installed(self, package_name, dnf_output):
+    def match_package_installed(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/dnf.py:230]: list vs. List

File: kiwi/package_manager/dnf.py [hunk 6]
```diff
@@ -242,10 +244,10 @@ def match_package_installed(self, package_name, dnf_output):
         :rtype: match object, None
         """
         return re.match(
-            '.*Installing  : ' + re.escape(package_name) + '.*', dnf_output
+            '.*Installing  : ' + re.escape(package_list) + '.*', log_line
         )
 
-    def match_package_deleted(self, package_name, dnf_output):
+    def match_package_deleted(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/dnf.py:250]: list vs. List

File: kiwi/package_manager/microdnf.py [hunk 7]
```diff
@@ -39,7 +42,7 @@ class PackageManagerMicroDnf(PackageManagerBase):
     :param dict command_env: microdnf command environment from repository runtime
         configuration
     """
-    def post_init(self, custom_args=None):
+    def post_init(self, custom_args: list = None) -> None:
```
Comment (schaefi) [inline kiwi/package_manager/microdnf.py:45]: list vs. List

File: kiwi/package_manager/microdnf.py [hunk 8]
```diff
@@ -209,21 +212,21 @@ def update(self):
...
-    def process_plus_recommended(self):
+    def process_plus_recommended(self) -> None:
         """
         Setup package processing to also include recommended dependencies.
         """
         if '--setopt=install_weak_deps=0' in self.custom_args:
             self.custom_args.remove('--setopt=install_weak_deps=0')
 
-    def match_package_installed(self, package_name, dnf_output):
+    def match_package_installed(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/microdnf.py:229]: list vs. List

File: kiwi/package_manager/microdnf.py [hunk 9]
```diff
@@ -240,10 +243,10 @@ def match_package_installed(self, package_name, dnf_output):
         :rtype: match object, None
         """
         return re.match(
-            '.*Installing  : ' + re.escape(package_name) + '.*', dnf_output
+            '.*Installing  : ' + re.escape(package_list) + '.*', log_line
         )
 
-    def match_package_deleted(self, package_name, dnf_output):
+    def match_package_deleted(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/microdnf.py:249]: list vs. List

File: kiwi/package_manager/pacman.py [hunk 10]
```diff
@@ -37,7 +39,7 @@ class PackageManagerPacman(PackageManagerBase):
     :param dict command_env: pacman command environment from repository
         runtime configuration
     """
-    def post_init(self, custom_args=None):
+    def post_init(self, custom_args: list = None) -> None:
```
Comment (schaefi) [inline kiwi/package_manager/pacman.py:42]: list vs. List

File: kiwi/package_manager/pacman.py [hunk 11]
```diff
@@ -194,23 +196,23 @@ def update(self):
...
+    def process_plus_recommended(self) -> None:
         """
         Setup package processing to also include recommended dependencies.
 
         There is no such concept of recommended dependencies in pacman.
         """
         pass
 
-    def match_package_installed(self, package_name, pacman_output):
+    def match_package_installed(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/pacman.py:215]: list vs. List

File: kiwi/package_manager/pacman.py [hunk 12]
```diff
@@ -227,10 +229,10 @@ def match_package_installed(self, package_name, pacman_output):
         :rtype: match object, None
         """
         return re.match(
-            '.* installing ' + re.escape(package_name) + '.*', pacman_output
+            '.* installing ' + re.escape(package_list) + '.*', log_line
         )
 
-    def match_package_deleted(self, package_name, pacman_output):
+    def match_package_deleted(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/pacman.py:235]: list vs. List

File: kiwi/package_manager/zypper.py [hunk 13]
```diff
@@ -38,7 +41,7 @@ class PackageManagerZypper(PackageManagerBase):
     :param dict command_env: zypper command environment from repository
         runtime configuration
     """
-    def post_init(self, custom_args=None):
+    def post_init(self, custom_args: list = None) -> None:
```
Comment (schaefi) [inline kiwi/package_manager/zypper.py:44]: list vs. List

File: kiwi/package_manager/zypper.py [hunk 14]
```diff
@@ -202,21 +205,21 @@ def update(self):
...
-    def process_plus_recommended(self):
+    def process_plus_recommended(self) -> None:
         """
         Setup package processing to also include recommended dependencies.
         """
         if '--no-recommends' in self.custom_args:
             self.custom_args.remove('--no-recommends')
 
-    def match_package_installed(self, package_name, zypper_output):
+    def match_package_installed(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/zypper.py:222]: list vs. List

File: kiwi/package_manager/zypper.py [hunk 15]
```diff
@@ -233,10 +236,10 @@ def match_package_installed(self, package_name, zypper_output):
         :rtype: match object, None
         """
         return re.match(
-            '.*Installing: ' + re.escape(package_name) + '.*', zypper_output
+            '.*Installing: ' + re.escape(package_list) + '.*', log_line
         )
 
-    def match_package_deleted(self, package_name, zypper_output):
+    def match_package_deleted(self, package_list: list, log_line: str) -> Union[Match, None]:
```
Comment (schaefi) [inline kiwi/package_manager/zypper.py:242]: list vs. List

**Review by schaefi — CHANGES_REQUESTED** [review]
Thanks, this is a great start :+1: 

See my comments, mostly nitpick but there is also one inconsistency which I think was handled incorrectly. More details in an extra comment

**Comment by schaefi** [comment]
@KeithMnemonic I saw you found an inconsistency in the package manager code for the match_package_installed and match_package_deleted functions. Great catch :+1: I however think it was solved in the wrong direction. If you look at ```kiwi/command_process.py``` you see how those methods are called. There is poll_show_progress() which iterates over a list of items (usually package names) and calls for each of them _update_progress(). In that update method you see

```python
match_method(item, command_output)
```

item is always a string and command_output is as well.

Thus the wrong information in the match_package_* methods is the comment block and not the parameters. Therefore I think the type hints should be changed to for example:

```python
def match_package_installed(self, package_name: str, zypper_output: str):
```

and the comment blocks in that functions that tells something about ```:param list package_list: list of all packages``` needs to be fixed, that is a wrong information from my perspective

Thanks much

**Comment by KeithMnemonic** [comment]
> @KeithMnemonic I saw you found an inconsistency in the package manager code for the match_package_installed and match_package_deleted functions. Great catch 👍 I however think it was solved in the wrong direction. If you look at `kiwi/command_process.py` you see how those methods are called. There is poll_show_progress() which iterates over a list of items (usually package names) and calls for each of them _update_progress(). In that update method you see
> 
> ```python
> match_method(item, command_output)
> ```
> 
> item is always a string and command_output is as well.
> 
> Thus the wrong information in the match_package_* methods is the comment block and not the parameters. Therefore I think the type hints should be changed to for example:
> 
> ```python
> def match_package_installed(self, package_name: str, zypper_output: str):
> ```
> 
> and the comment blocks in that functions that tells something about `:param list package_list: list of all packages` needs to be fixed, that is a wrong information from my perspective
> 
> Thanks much

Markus, I can change it back to package_name and fix the comments block. I am not sure zypper_output will work as it seems the "Codacy" check was complaining and failing about renaming the parameters from the base class base.py and the individual subclasses such as dnf.py,  zypper.py

See this

https://app.codacy.com/gh/OSInside/kiwi/pullRequest?prid=7088745

and let me know how to get around it

**Comment by schaefi** [comment]
>the "Codacy" check was complaining and failing about renaming the parameters from the base class base.py and the individual subclasses such as dnf.py, zypper.py

yes, it's good that this gets cleaned up now. My suggestion would be. In the base class

```python
def match_package_installed(self, package_name: str, package_manager_output: str) -> bool:
        """
        Match expression to indicate a package has been installed

        Implementation in specialized package manager class

        :param str package_name: unused
        :param str package_manager_output: unused
        """
        raise NotImplementedError
```

and the implementations for the package managers follows this. I'm not sure how mypy reacts on the return statement.
Actually the best solution would be to refactor the base class into a virtual class using ```from abc import ABCMeta, abstractmethod``` But I think this should be an extra pull request as it imho would be too much for the scope of adding type hints in the first place

The other error from codacy ```Method could be a function``` can be solved by using the ```@staticmethod``` decorator

```python
@staticmethod
def has_failed(returncode):
    ...
```

Does this help ?

**Comment by KeithMnemonic** [comment]
> > the "Codacy" check was complaining and failing about renaming the parameters from the base class base.py and the individual subclasses such as dnf.py, zypper.py
> 
> yes, it's good that this gets cleaned up now. My suggestion would be. In the base class
> 
> ```python
> def match_package_installed(self, package_name: str, package_manager_output: str) -> bool:
>     """
>     Match expression to indicate a package has been installed
> 
>     Implementation in specialized package manager class
> 
>     :param str package_name: unused
>     :param str package_manager_output: unused
>     """
>     raise NotImplementedError
> ```
> 
> and the implementations for the package managers follows this. I'm not sure how mypy reacts on the return statement.
> Actually the best solution would be to refactor the base class into a virtual class using `from abc import ABCMeta, abstractmethod` But I think this should be an extra pull request as it imho would be too much for the scope of adding type hints in the first place
> 
> The other error from codacy `Method could be a function` can be solved by using the `@staticmethod` decorator
> 
> ```python
> @staticmethod
> def has_failed(returncode):
>     ...
> ```
> 
> Does this help ?

Marcus,

from the earlier mypy and codacy failurees, if we are using 

return re.match(
            '.*Unpacking ' + re.escape(package_list) + '.*', log_line
        )

I do  not think ->bool will work in the base class per 

re.match(pattern, string, flags=0)
If zero or more characters at the beginning of string match the regular expression pattern, return a corresponding match object. Return None if the string does not match the pattern; note that this is different from a zero-length match.

that is why I needed the union. Do you want to change all of the subclasses to return True if 'match' and False if 'None'?

**Comment by schaefi** [comment]
>I do not think ->bool will work in the base class per

>re.match(pattern, string, flags=0)
>If zero or more characters at the beginning of string match the regular expression pattern, return a corresponding match object. >Return None if the string does not match the pattern; note that this is different from a zero-length match.

>that is why I needed the union. Do you want to change all of the subclasses to return True if 'match' and False if 'None'?

I agree with you. I think the code should change to return a bool value. From my perspective the bool operator in python should do the trick

```python
return bool(
    re.match('...', package_manager_output)
)
```

Thanks

File: kiwi/package_manager/apt.py (hunk 2 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/apt.py]: fixed in latest push

File: kiwi/package_manager/dnf.py (hunk 4 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/dnf.py]: fixed in latest push

File: kiwi/package_manager/pacman.py (hunk 11 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/pacman.py]: fixed in latest push

File: kiwi/package_manager/zypper.py (hunk 13 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/zypper.py:44]: fixed in latest push

File: kiwi/package_manager/dnf.py (hunk 6 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/dnf.py]: fixed in latest push

File: kiwi/package_manager/microdnf.py (hunk 8 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/microdnf.py]: fixed in latest push

File: kiwi/package_manager/pacman.py (hunk 10 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/pacman.py]: fixed in latest push

File: kiwi/package_manager/microdnf.py (hunk 9 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/microdnf.py]: fixed in latest push

File: kiwi/package_manager/microdnf.py (hunk 7 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/microdnf.py]: fixed in latest push

**Comment by KeithMnemonic** [comment]
> > I do not think ->bool will work in the base class per
> 
> > re.match(pattern, string, flags=0)
> > If zero or more characters at the beginning of string match the regular expression pattern, return a corresponding match object. >Return None if the string does not match the pattern; note that this is different from a zero-length match.
> 
> > that is why I needed the union. Do you want to change all of the subclasses to return True if 'match' and False if 'None'?
> 
> I agree with you. I think the code should change to return a bool value. From my perspective the bool operator in python should do the trick
> 
> ```python
> return bool(
>     re.match('...', package_manager_output)
> )
> ```
> 
> Thanks

Markus, please re-review and let me know if this looks good now.

File: kiwi/package_manager/base.py [hunk 16]
```diff
@@ -124,73 +129,82 @@ def process_delete_requests(self, force=False):
...
         Match expression to indicate a package has been installed
 
         Implementation in specialized package manager class
 
-        :param list package_list: unused
-        :param str log_line: unused
+        :param str package_name: unused
+        :param str package_manager_output: unused
+
+        :return: Match|None
```
Comment (schaefi) [inline kiwi/package_manager/base.py:165]: I think that should be better ```:return: True|False```

File: kiwi/package_manager/base.py [hunk 17]
```diff
@@ -124,73 +129,82 @@ def process_delete_requests(self, force=False):
...
+
+        :rtype: Match or None
```
Comment (schaefi) [inline kiwi/package_manager/base.py:167]: return type is ```:rtype: bool```

**Review by schaefi — COMMENTED** [review]
That looks very good now

just nitpick on the doc string describing the return and rtype values. I think after you changed that to return bool(...) this information should be adapted in all methods that are affected

File: kiwi/package_manager/base.py (hunk 16 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/base.py]: please check again with latest push

File: kiwi/package_manager/base.py (hunk 17 above)
Comment (KeithMnemonic) [reply kiwi/package_manager/base.py]: please check again with lastest push

**Review by schaefi — APPROVED** [review]

**Pull request merged**